[JIT] additional support for CallMethod with autocasting

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #67925

Previously, the following would always fail, because autocasting would not be enabled in the called method:

```
@torch.jit.script
def fn(x, y):
    with autocast():
        # CallMethod() to some method

fn(x, y)
```

This allows the above, if autocasting is globally enabled, e.g.

```
@torch.jit.script
def fn(x, y):
    with autocast():
        # CallMethod() to some method

with autocast():
    fn(x, y) # now
```

Differential Revision: [D32214439](https://our.internmc.facebook.com/intern/diff/D32214439/)

davidberard98 committed:
[JIT] additional support for CallMethod with autocasting

Previously, the following would always fail, because autocasting would not be enabled in the called method:

```
@torch.jit.script
def fn(x, y):
    with autocast():
        # CallMethod() to some method

fn(x, y)
```

This allows the above, if autocasting is globally enabled, e.g.

```
@torch.jit.script
def fn(x, y):
    with autocast():
        # CallMethod() to some method

with autocast():
    fn(x, y) # now
```

Differential Revision: [D32214439](https://our.internmc.facebook.com/intern/diff/D32214439/)

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] additional support for CallMethod with autocasting"


Previously, the following would always fail, because autocasting would not be enabled in the called method:

```
torch.jit.script
def fn(x, y):
    with autocast():
        # CallMethod() to some method

fn(x, y)
```

This allows the above, if autocasting is globally enabled, e.g.

```
torch.jit.script
def fn(x, y):
    with autocast():
        # CallMethod() to some method

with autocast():
    fn(x, y) # now
```

Differential Revision: [D32214439](https://our.internmc.facebook.com/intern/diff/D32214439/)

[ghstack-poisoned]

eellison commented on torch/csrc/jit/passes/autocast.cpp:
```diff
@@ -446,9 +457,12 @@ bool autocastEnabled() {
-        at::autocast::is_cpu_enabled(),
+        is_gpu_enabled,
+        is_cpu_enabled,
+        is_gpu_enabled || is_cpu_enabled,
```
This should just be always true, since we are reflecting if global state is set appropriately, not whether it is set to true

davidberard98 committed:
Update on "[JIT] additional support for CallMethod with autocasting"


Previously, the following would always fail, because autocasting would not be enabled in the called method:

```
torch.jit.script
def fn(x, y):
    with autocast():
        # CallMethod() to some method

fn(x, y)
```

This allows the above, if autocasting is globally enabled, e.g.

```
torch.jit.script
def fn(x, y):
    with autocast():
        # CallMethod() to some method

with autocast():
    fn(x, y) # now
```

Differential Revision: [D32214439](https://our.internmc.facebook.com/intern/diff/D32214439/)

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] additional support for CallMethod with autocasting"


Previously, the following would always fail, because autocasting would not be enabled in the called method:

```
torch.jit.script
def fn(x, y):
    with autocast():
        # CallMethod() to some method

fn(x, y)
```

This allows the above, if autocasting is globally enabled, e.g.

```
torch.jit.script
def fn(x, y):
    with autocast():
        # CallMethod() to some method

with autocast():
    fn(x, y) # now
```

Differential Revision: [D32214439](https://our.internmc.facebook.com/intern/diff/D32214439/)

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] additional support for CallMethod with autocasting"


Previously, the following would always fail, because autocasting would not be enabled in the called method:

```
torch.jit.script
def fn(x, y):
    with autocast():
        # CallMethod() to some method

fn(x, y)
```

This allows the above, if autocasting is globally enabled, e.g.

```
torch.jit.script
def fn(x, y):
    with autocast():
        # CallMethod() to some method

with autocast():
    fn(x, y) # now
```

Differential Revision: [D32214439](https://our.internmc.facebook.com/intern/diff/D32214439/)

[ghstack-poisoned]

eellison reviewed (approved):
lgtm
